liblangutil/SourceLocation.h

```diff
@@ -67,37 +68,46 @@ struct SourceLocation
 
 	bool isEmpty() const { return start == -1 && end == -1; }
 
+	bool hasSource() const noexcept { return charStream; }
+	bool hasSourceName() const noexcept { return charStream && !charStream->filename().empty(); }
+
+	std::string const& sourceName() const
+	{
+		static std::string const empty;
```

Comment:
This feels inconsistent with regards to `hasSourceName`.

Reply:
I'm open for suggestions. I did not want to return a `shared_ptr` or const pointer. that's a common trick I've been used to (although, not fixated to) -> boost::optional<std::string> would be the better fit then, IMHO.